demos/quic/server/server.c

```diff
@@ -0,0 +1,230 @@
+#include <openssl/err.h>
+#include <openssl/ssl.h>
+#include <openssl/quic.h>
+#include <netinet/in.h>
+#include <unistd.h>
+#include <assert.h>
+
+/*
+ * This is a basic demo of QUIC server functionality in which one connection at
+ * a time is accepted in a blocking loop.
+ */
+
+/* ALPN string for TLS handshake */
+static const unsigned char alpn_ossltest[] = {
+    /* "\x08ossltest" (hex for EBCDIC resilience) */
+    0x08, 0x6f, 0x73, 0x73, 0x6c, 0x74, 0x65, 0x73, 0x74
+};
+
+/* This callback validates and negotiates the desired ALPN on the server side. */
+static int select_alpn(SSL *ssl,
+                       const unsigned char **out, unsigned char *out_len,
+                       const unsigned char *in, unsigned int in_len,
+                       void *arg)
+{
+    if (SSL_select_next_proto((unsigned char **)out, out_len,
+                              alpn_ossltest, sizeof(alpn_ossltest), in, in_len)
+            != OPENSSL_NPN_NEGOTIATED)
+        return SSL_TLSEXT_ERR_ALERT_FATAL;
+
+    return SSL_TLSEXT_ERR_OK;
+}
+
+/* Create SSL_CTX. */
+static SSL_CTX *create_ctx(const char *cert_path, const char *key_path)
+{
+    SSL_CTX *ctx;
+
+    ctx = SSL_CTX_new(OSSL_QUIC_server_method());
+    if (ctx == NULL)
+        goto err;
+
+    /* Load certificate and corresponding private key. */
+    if (SSL_CTX_use_certificate_chain_file(ctx, cert_path) <= 0) {
+        fprintf(stderr, "couldn't load certificate file: %s\n", cert_path);
+        goto err;
+    }
+
+    if (SSL_CTX_use_PrivateKey_file(ctx, key_path, SSL_FILETYPE_PEM) <= 0) {
+        fprintf(stderr, "couldn't load key file: %s\n", key_path);
+        goto err;
+    }
+
+    if (!SSL_CTX_check_private_key(ctx)) {
+        fprintf(stderr, "private key check failed\n");
+        goto err;
+    }
+
+    /* Setup ALPN negotiation callback. */
+    SSL_CTX_set_alpn_select_cb(ctx, select_alpn, NULL);
+    return ctx;
+
+err:
+    SSL_CTX_free(ctx);
+    return NULL;
+}
+
+/* Create UDP socket using given port. */
+static int create_socket(uint16_t port)
+{
+    int fd = -1;
+    struct sockaddr_in6 sa = {0};
+
+    if ((fd = socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP)) < 0) {
```

Comment:
An IPv6 only demo?

Reply:
Most OSes enable v4-mapped addresses by default so this is dual-stack. There are some exceptions like OpenBSD &mdash; but handling those correctly would require multiple sockets here, significantly increasing the complexity of the demo and reducing its expository value.

I'm not massively bothered either way though.

Reply:
Windows doesn't work that way ... you need two listeners to get both IPv4 and IPv6 support and as that is one of the primary platforms making this choice here wasn't what I was expecting.

Reply:
It does support it from Vista onwards, though you do have to enable it explicitly.

Anyway, it doesn't really matter either way. AF_INET works for me.